crates/bevy_ecs/src/schedule/set.rs

```diff
@@ -5,61 +5,147 @@ use std::marker::PhantomData;
-pub type BoxedSystemSet = Box<dyn SystemSet>;
-/// A shorthand for `Box<dyn ScheduleLabel>`.
-pub type BoxedScheduleLabel = Box<dyn ScheduleLabel>;
-
-/// Types that identify logical groups of systems.
+/// Identifies a logical group of systems.
 pub trait SystemSet: DynHash + Debug + Send + Sync + 'static {
-    /// Returns `Some` if this system set is a [`SystemTypeSet`].
+    /// Returns the [`TypeId`] of the system if the system set is a [`SystemTypeSet`].
+    ///
+    /// A [`SystemTypeSet`] has special properties:
+    /// - You cannot manually add systems or sets to it.
+    /// - You cannot configure it.
+    /// - You cannot order relative to it if it contains more than one instance.
+    ///
+    /// These sets are automatically populated, so these constraints exist to prevent unintentional ambiguity.
     fn system_type(&self) -> Option<TypeId> {
         None
     }
 
-    /// Returns `true` if this system set is an [`AnonymousSet`].
+    /// Returns `true` if the system set is an [`AnonymousSet`].
     fn is_anonymous(&self) -> bool {
         false
     }
-    /// Creates a boxed clone of the label corresponding to this system set.
-    fn dyn_clone(&self) -> Box<dyn SystemSet>;
+
+    /// Returns the unique, type-elided identifier for the system set.
+    fn as_label(&self) -> SystemSetId {
+        SystemSetId::of(self)
+    }
+
+    /// Returns the type-elided version of the system set.
+    fn as_untyped(&self) -> SystemSetUntyped {
+        SystemSetUntyped::of(self)
+    }
 }
 
-impl PartialEq for dyn SystemSet {
-    fn eq(&self, other: &Self) -> bool {
-        self.dyn_eq(other.as_dyn_eq())
+/// A lightweight and printable identifier for a [`SystemSet`].
+#[derive(Clone, Copy, Eq)]
+pub struct SystemSetId(&'static str);
+
+impl SystemSetId {
+    /// Returns the [`SystemSetId`] of the [`SystemSet`].
+    pub fn of<S: SystemSet + ?Sized>(set: &S) -> SystemSetId {
+        let str = bevy_utils::label::intern_debug_string(set);
+        SystemSetId(str)
     }
 }
 
-impl Eq for dyn SystemSet {}
+impl PartialEq for SystemSetId {
+    fn eq(&self, other: &Self) -> bool {
+        std::ptr::eq(self.0, other.0)
+    }
+}
 
-impl Hash for dyn SystemSet {
+impl Hash for SystemSetId {
     fn hash<H: Hasher>(&self, state: &mut H) {
-        self.dyn_hash(state);
+        std::ptr::hash(self.0, state);
     }
 }
 
-impl Clone for Box<dyn SystemSet> {
-    fn clone(&self) -> Self {
-        self.dyn_clone()
+impl Debug for SystemSetId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+/// The different variants of system sets.
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+enum SystemSetKind {
+    Anonymous,
+    Named,
+    SystemType(TypeId),
+}
+
+/// Type-elided struct whose methods return the same values as its original [`SystemSet`].
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+
+pub struct SystemSetUntyped {
+    id: SystemSetId,
+    kind: SystemSetKind,
+}
```

Comment:
I think the presence of `SystemSetId` and `SystemSetUntyped` is a little confusing -- could this types be merged? (Of course, doing this would probably mean you can't use `define_label!` for `SystemSet`).

Reply:
`SystemSetId` is just the ID. To replace the boxing in schedule construction, we also need to remember the distinction between normal system sets, system type sets, and anonymous system sets.

Reply:
Right, but is it necessary to have `SystemSetId` as a separate public type? Can we just use `SystemSetUntyped` for any user-facing APIs?

Reply:
Hmm, yeah, we can probably keep `SystemSetId` internal.